exonum-java-binding-core/src/test/java/com/exonum/binding/storage/indices/BaseListIndexIntegrationTestable.java

```diff
@@ -301,6 +306,76 @@ void testIterator() {
     });
   }
 
+  @Test
+  void testStream() {
+    runTestWithView(database::createFork, (l) -> {
+      List<String> elements = TestStorageItems.values;
+
+      l.addAll(elements);
+
+      List<String> streamElements = l.stream()
+          .collect(Collectors.toList());
+
+      assertThat(streamElements, equalTo(elements));
+    });
+  }
+
+  @Test
+  void streamIsLateBinding() {
+    runTestWithView(database::createFork, (l) -> {
+      // Add some elements
+      l.add(V1);
+      l.add(V2);
+
+      // Create a stream
+      Stream<String> stream = l.stream();
+
+      // Replace the first
+      l.set(0, V3);
+
+      // Check the updated values are available, with no exceptions
+      String[] streamElements = stream.toArray(String[]::new);
+      assertThat(streamElements, arrayContaining(V3, V2));
+    });
+  }
+
+  @Test
+  void streamIsLateBindingInCaseOfStructuralModifications() {
+    runTestWithView(database::createFork, (l) -> {
+      // Add an element
+      l.add(V1);
+
+      // Create a stream
+      Stream<String> stream = l.stream();
+
+      // Add a new one
+      l.add(V2);
+
+      // Check the updated values are available, with no exceptions
+      String[] streamElements = stream.toArray(String[]::new);
+      assertThat(streamElements, arrayContaining(V1, V2));
+    });
+  }
+
+  @Test
+  void streamDetectsStructuralModifications() {
+    runTestWithView(database::createFork, (l) -> {
+      // Add some elements
+      l.add(V1);
+      l.add(V2);
+
+      // Create a stream that modifies the collection while it is evaluated
+      long maxIterations = l.size();
+      assertThrows(ConcurrentModificationException.class, () -> l.stream()
+          // Perform a structurally-interfering operation on the source list
+          .peek(l::add)
+          // Limit, so that we don't get infinite number of iterations in case of a broken
+          // spliterator
+          .limit(maxIterations)
+          .count());
+    });
+  }
+
```

Comment:
Do we need to Test `stream().parallel()...`?

Reply:
I think no, since we do not currently support that — all proxies are documented to be non-thread-safe (but it is likely work).